Rescue JSON decoding errors rather than killing WorkerManager

keyan opened the pull request:
This is a first pass at addressing https://github.com/edgurgel/verk/issues/89. I was initially unsure of how to skip jobs because `QueueManager.dequeue`will have already moved them to the `inprogress` queue. But calling `QueueManager.ack` will remove it.

It might be cleaner to add a new handler to `QueueManager` that does the same thing as `:ack` but is explicitly used for removal (whereas using `:ack` seems to imply removal of a successfully completed job).

It is pretty difficult to add a test for this as enqueuing jobs from Verk uses Poison encoding and thus the interoperability problem I am running into cannot be recreated easily.

mitchellhenke commented on lib/verk/workers_manager.ex:
```diff
@@ -99,7 +99,15 @@ defmodule Verk.WorkersManager do
-          for job <- jobs, do: job |> Job.decode! |> start_job(state)
+          Enum.map(jobs, fn job ->
+            try do
+              job |> Job.decode! |> start_job(state)
```
Would it be preferable to have `Job.decode!` not raise, and instead return `{:ok, job}`, `{:error reason}` tuple or at least rescue on a specific exception type?

edgurgel replied:
Yeah I think we should use `Job.decode` -> `Poison.decode` and treat the error case 👍

keyan replied:
👍 okay that sounds good, and treat by `ack`ing then sending to the dead set?

edgurgel replied:
I'm fine either by just logging or adding to the dead set. We just need to add to the dead set before `ack` so that we don't have a ghost job:

For example:

* Ack!
* Add dead job (fails) (BOOM)

The job is gone from progress and it wasn't added to the dead set.

If we go to the Dead Set route we could potentially use the original json and add a clause for the job being already a string? so DeadSet.add would have an implementation if a Job is sent and one if a string is sent?

edgurgel replied:
But another problem may come having it inside the Dead set: listing the dead jobs won't be as easy as we don't know if they are decodable so Verk Web for example can't show all the jobs etc

mitchellhenke commented on lib/verk/workers_manager.ex:
```diff
@@ -99,7 +99,15 @@ defmodule Verk.WorkersManager do
       case QueueManager.dequeue(state.queue_manager_name, workers) do
         jobs when is_list(jobs) ->
-          for job <- jobs, do: job |> Job.decode! |> start_job(state)
+          Enum.map(jobs, fn job ->
```
Could this be `Enum.each`?

keyan commented:
@edgurgel, ready for another look.

After taking into consideration both of your suggestions, I realized the diff was quite large, so I rebased. Also added some tests for the new `Job.decode/1` and updated `WorkerManager.handle_info/2` methods.

keyan commented:
I just ran into an issue with this fix, when we try to remove the job from the `inprogress` queue by calling `QueueManager.ack/2` the `handle_cast` callback attempts to [access `job.original_json`](https://github.com/edgurgel/verk/blob/ddd67707f65d2444115c460f56b932fa5b9a4a8e/lib/verk/queue_manager.ex#L147), which raises:

```
18:36:32.178 [error] GenServer :"php_queue:emails.queue_manager" terminating
** (ArgumentError) argument error
    :erlang.apply("", :original_json, [])
    (verk) Verk.QueueManager.handle_cast/2
    (stdlib) gen_server.erl:601: :gen_server.try_dispatch/4
    (stdlib) gen_server.erl:667: :gen_server.handle_msg/5
    (stdlib) proc_lib.erl:247: :proc_lib.init_p_do_apply/3
Last message: {:"$gen_cast", {:ack, ""}}
State: %Verk.QueueManager.State{node_id: "1", queue_name: :"php_queue:emails", redis: #PID<0.5955.0>}
```

Does it make sense to have a new handler in `QueueManager`? Say:

```elixir
def handle_cast({:malformed, job}, state) do
  case Redix.command(state.redis, ["LREM", inprogress(state.queue_name, state.node_id), "-1", job]) do
    {:ok, 1} -> :ok
    _ -> Logger.error("Failed to acknowledge job #{inspect job}")
  end
  {:noreply, state}
end
```

edgurgel commented:
Hey sorry for not replying but this week has been pretty busy. I'm keen to move this forward this weekend. Thank you again for pushing this forward!

keyan commented:
No worries! Thanks for staying on top of this.

We've been running this branch in production for the last couple days without any issues. Of course I am happy to work on this further to make sure we can get a fix merged so we can stop using a fork in production 😄 .

edgurgel commented:
LGTM! I will release a new version soon! 🎉
